sign.go

```diff
@@ -1,139 +1,181 @@
-		w.Write([]byte(strings.ToLower(h)))
-		w.Write([]byte(":"))
-		w.Write([]byte(strings.Join(v, ",")))
-		w.Write([]byte("\n"))
+	sort.Strings(headers)
+
+	s.signedHeaders = strings.Join(headers, ";")
+
+	headerValues := make([]string, len(headers))
+	for i, k := range headers {
+		if k == "host" {
+			headerValues[i] = "host:" + s.Request.URL.Host
+		} else {
+			headerValues[i] = k + ":" +
+				strings.Join(s.Request.Header[http.CanonicalHeaderKey(k)], ",")
+		}
 	}
+
+	s.canonicalHeaders = strings.Join(headerValues, "\n")
 }
 
-// From Amazon API documentation:
-//
-// CanonicalizedResource = [ "/" + Bucket ] +
-//    <HTTP-Request-URI, from the protocol name up to the query string> +
-//    [ subresource, if present. For example "?acl", "?location", "?logging", or "?torrent"];
-func (b *Bucket) writeCanonicializedResource(w io.Writer, r *http.Request) {
-	if !strings.Contains(b.Name, ".") { // handling for bucket names containing periods
-		w.Write([]byte("/"))
-		w.Write([]byte(b.Name))
+func (s *signer) buildCanonicalString() {
+	s.Request.URL.RawQuery = strings.Replace(s.Request.URL.Query().Encode(), "+", "%20", -1)
+	uri := s.Request.URL.Opaque
+	if uri != "" {
+		uri = "/" + strings.Join(strings.Split(uri, "/")[3:], "/")
+	} else {
+		uri = s.Request.URL.EscapedPath()
 	}
-	u := &url.URL{Path: r.URL.Path}
-	w.Write([]byte(u.String()))
-	b.writeSubResource(w, r)
+	if uri == "" {
+		uri = "/"
+	}
+
+	s.canonicalString = strings.Join([]string{
+		s.Request.Method,
+		uri,
+		s.Request.URL.RawQuery,
+		s.canonicalHeaders + "\n",
+		s.signedHeaders,
+		s.bodyDigest(),
+	}, "\n")
 }
 
-// See Amazon S3 Developer Guide for explanation
-// http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html
-func (b *Bucket) writeSubResource(w io.Writer, r *http.Request) {
-	var sr []string
-	for k, vs := range r.URL.Query() {
-		if paramsToSign[k] {
-			for _, v := range vs {
-				if v == "" {
-					sr = append(sr, k)
-				} else {
-					sr = append(sr, k+"="+v)
-				}
-			}
+func (s *signer) buildStringToSign() {
+	s.stringToSign = strings.Join([]string{
+		prefix,
+		s.Time.UTC().Format(isoFormat),
+		s.credentialString,
+		hex.EncodeToString(sha([]byte(s.canonicalString))),
+	}, "\n")
+}
+
+func (s *signer) buildSignature() {
+	secret := s.Keys.SecretKey
+	date := hmacSign([]byte("AWS4"+secret), []byte(s.Time.UTC().Format(shortDate)))
+	region := hmacSign(date, []byte(s.Region))
+	service := hmacSign(region, []byte("s3"))
+	credentials := hmacSign(service, []byte("aws4_request"))
+	signature := hmacSign(credentials, []byte(s.stringToSign))
+	s.signature = hex.EncodeToString(signature)
+}
+
+func (s *signer) bodyDigest() string {
+	hash := s.Request.Header.Get("X-Amz-Content-Sha256")
+	if hash == "" {
+		if s.Request.Body == nil {
+			hash = hex.EncodeToString(sha([]byte{}))
+		} else {
+			body, _ := ioutil.ReadAll(s.Request.Body)
+			s.Request.Body = ioutil.NopCloser(bytes.NewReader(body))
+			hash = hex.EncodeToString(sha(body))
```

Comment:
You're right, this is a ton of copies and will significantly increase memory usage for large transfers. 

As you suggested, the `X-Amz-Content-Sha256` header can be calculated in putter and getter when the body is seekable.

I'd like to get that change in before merging this to avoid a huge efficiency regression.